cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/oidc/OIDCUtil.java

```diff
@@ -0,0 +1,202 @@
+ *
+ *  You should have received             Vous devriez avoir reçu une
+ *  a copy of the GNU Affero             copie de la Licence Générale
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ ************************************************************************
+ */
+package ca.nrc.cadc.ac.server.oidc;
+
+import ca.nrc.cadc.ac.Group;
+import ca.nrc.cadc.ac.Role;
+import ca.nrc.cadc.ac.User;
+import ca.nrc.cadc.ac.UserNotFoundException;
+import ca.nrc.cadc.ac.server.GroupPersistence;
+import ca.nrc.cadc.ac.server.UserPersistence;
+import ca.nrc.cadc.ac.server.ldap.LdapGroupPersistence;
+import ca.nrc.cadc.ac.server.ldap.LdapUserPersistence;
+import ca.nrc.cadc.auth.AuthenticationUtil;
+import ca.nrc.cadc.auth.HttpPrincipal;
+import ca.nrc.cadc.auth.NumericPrincipal;
+import ca.nrc.cadc.auth.SignedToken;
+import ca.nrc.cadc.net.TransientException;
+
+import java.io.IOException;
+import java.net.URI;
+import java.security.AccessControlException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.KeyPair;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.security.auth.Subject;
+
+import org.apache.log4j.Logger;
+
+import io.jsonwebtoken.JwtBuilder;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.jsonwebtoken.security.Keys;
+
+/**
+ * 
+ * Utilities for OIDC/OAuth2 support.
+ * 
+ * @author majorb
+ *
+ */
+public class OIDCUtil {
+    
+    public static final String AUTHORIZE_TOKEN_SCOPE = "cadc:oauth2/authorize_token";
+    public static final String REFRESH_TOKEN_SCOPE = "cadc:oauth2/refresh_token";
+    public static final String ACCESS_TOKEN_SCOPE = "cadc:oauth2/access_token";
+    
+    public static final Integer ID_TOKEN_EXPIRY_MINUTES = 60*24*7*2; // 2 weeks
+    public static final Integer AUTHORIZE_CODE_EXPIRY_MINUTES = 10;
+    public static final Integer ACCESS_CODE_EXPIRY_MINUTES = 60*24*7*2; // 2 weeks
+    public static final Integer REFRESH_TOKEN_EXPIRY_MINUTES = 60*24*7*52; // 1 year
+    public static final Integer JWT_EXPIRY_MINUTES = 60*24*7*2; // 2 weeks
+    
+    public static final String CLAIM_ISSUER_VALUE = "https://proto.canfar.net/ac";
+    public static final String CLAIM_GROUPS_KEY = "memberOf";
+    
+    static final Key publicSigningKey;
+    static final Key privateSigningKey;
+    
+    private static final Logger log = Logger.getLogger(OIDCUtil.class);
+    
+    private static final Map<String, RelyParty> relyParties;
+    
+    static {
+        // add all rely parties
+        relyParties = new HashMap<String, RelyParty>();
```

Comment:
Looks like a hard-coded RelyParty below. Should the value(s) come from config? It seems like the use of a map means you anticipate there being other RelyParty(s) in future. 

There are several ac config files already... ac-domains.properties has a single key in it now and seems like a "configuring the site" kind of file; could use it for now but plan on renaming it to ac.properties - having that single service config file has proven to be a good concept in the inventory work

Reply:
Yes, the idea is that we can add more rely parties (ie oidc clients) to that map.

Reply:
If you look in UserInfoAction, you'll see that we can't add a second rely party until we keep track of the client_id our tokens.  So, I suggest we do the work to allow multiple rely parties later.  I'll add a comment about the properties file use so we don't forget.